Clearly define discrete time variable (#2989)
modelica/ModelicaSpecification: chapters/classes.tex

\begin{itemize}
\item
  A variable \lstinline!vc! declared with \lstinline!constant!\indexinline{constant} prefix remains constant during transient analysis, with a value that is unaffected by the initialization problem.
  This is called a \firstuse{constant}, or \firstuse{constant!variable}\index{constant!variable}\index{component variability!constant}.
\item
  A variable \lstinline!vc! declared with the \lstinline!parameter!\indexinline{parameter} prefix remains constant during transient analysis, with a value determined by the initialization problem.
  This is called a \firstuse{parameter}, or \firstuse{parameter!variable}\index{parameter!variable}\index{component variability!parameter}.
\item

>   A \emph{discrete-time}\index{discrete-time!variable}\index{component variability!discrete-time} variable \lstinline!vd! is a variable that is discrete-valued (that is, not of \lstinline!Real! type) or assigned in a \lstinline!when!-clause.

Reviewer: What I was trying to suggest with `\firstuse` was simply this:
```suggestion
  A \firstuse{discrete-time variable}\index{discrete-time!variable}\index{component variability!discrete-time} \lstinline!vd! is a variable that is discrete-valued (that is, not of \lstinline!Real! type) or assigned in a \lstinline!when!-clause.
```

Author: Yes, but I read the comments and `firstuse` indicates that it later might automatically add the text to the index - and that will not work if the index should contain something slightly different as in this case. That's why I was hesitant to use it.

Reviewer: I see.  However note that:
- There is currently one or `\index` calls with every `\firstuse`.
- The argument to `\index` is often, but not always the same as the argument to `\firstuse`.

My conclusion is that the comment is making more harm than good.  If we'd ever change `\firstuse` to also add to the index, we'd have to go through all the places anyway and remove the current `\index` calls that would then become redundant.

I see two possibilities for changing the comment and/or implementation of `\firstuse`:
- Maybe we should change the comment to settle that `\firstuse` is all about semantic markup for styling, and explain in the comment that it should go together with `\index`, and that it is sometimes the case that the argument to `\index` is structured a bit differently that the argument to `\firstuse`?
- Alternatively, we could give `\firstuse` an optional argument for how the introduced terminology should be presented in the index, and by default just use the mandatory argument as-is.

How about going with `\firstuse` instead of `\emph` in this case just to follow the same pattern as elsewhere, and open a separate issue about sorting out what `\firstuse` really should do and say it might do in the future?

Reviewer: With the continuous-time variable item being fixed, this is now the only remaining unresolved comment.

  The \lstinline!discrete!\indexinline{discrete} prefix may be used to clarify that a variable is discrete-time.
  It has a vanishing time derivative between events.
  Note that this is not the same as saying that \lstinline!der(vd)=0! almost everywhere, as the derivative is not even defined at the events.
  It is not allowed to apply \lstinline!der! to discrete-time variables.
  During transient analysis the variable can only change its value at event instants (see \cref{events-and-synchronization}).
\item

>   A \emph{continuous-time}\index{continuous-time!variable}\index{component variability!continuous-time} variable is a \lstinline!Real! variable without any prefix that is not assigned in a \lstinline!when!-clause.

Reviewer: What I was trying to suggest with \firstuse was simply this:
```suggestion
  A \firstuse{continuous-time variable}\index{continuous-time!variable}\index{component variability!continuous-time} is a \lstinline!Real! variable without any prefix that is not assigned in a \lstinline!when!-clause.
```

>   A continuous-time variable \lstinline!vn! may have a non-vanishing time derivative (\lstinline!der(vn)<>0! possible - if \lstinline!der(vn)! is allowed) and may also change its value discontinuously at any time during transient analysis (see \cref{events-and-synchronization}).

Reviewer: En-dash, or rewrite with comma:
```suggestion
  A continuous-time variable \lstinline!vn! may have a non-vanishing time derivative (\lstinline!der(vn) <> 0! is possible, provided \lstinline!der(vn)! is defined) and may also change its value discontinuously at any time during transient analysis (see \cref{events-and-synchronization}).
```

  It may also contain a combination of these effects.

>   Note that \lstinline!der(vn)! is not allowed if there are any discontinuities, see \cref{modelica:der}.

Reviewer: Just a thought (not really requesting change): Do we really want to repeat this rather problematic statement in view of variables affected by `\reinit`?  Maybe localizing the problem to the definition of `der` would be preferable?

This variant would avoid repeating the problematic part:
```suggestion
  Regarding existence of \lstinline!der(vn)!, see \cref{modelica:der}.
```

\begin{itemize}
\item
  A variable \lstinline!vc! declared with \lstinline!constant!\indexinline{constant} prefix remains constant during transient analysis, with a value that is unaffected by the initialization problem.
  This is called a \firstuse{constant}, or \firstuse{constant variable}\index{constant variable}\index{component variability!constant}.
\item
  A variable \lstinline!vc! declared with the \lstinline!parameter!\indexinline{parameter} prefix remains constant during transient analysis, with a value determined by the initialization problem.
  This is called a \firstuse{parameter}, or \firstuse{parameter variable}\index{parameter variable}\index{component variability!parameter}.
\item

>   A \emph{discrete-time} variable \lstinline!vd! is a variable that is discrete-valued (that is, not of \lstinline!Real! type) or assigned in a \lstinline!when!-clause.

Reviewer: Should probably be introduced more similar to constants and parameters, something like this:
```suggestion
  A \firstuse{discrete-time variable}\index{discrete-time variable}\index{component variability!discrete-time} \lstinline!vd! is a variable that is discrete-valued (that is, not of \lstinline!Real! type) or assigned in a \lstinline!when!-clause.
```

  The \lstinline!discrete!\indexinline{discrete} prefix may be used to clarify that a variable is discrete-time.
  It has a vanishing time derivative between events.
  Note that this is not the same as saying that \lstinline!der(vd)=0! almost everywhere, as the derivative is not even defined at the events.
  It is not allowed to apply \lstinline!der! to discrete-time variables.
  During transient analysis the variable can only change its value at event instants (see \cref{events-and-synchronization}).
\item
  A \emph{continuous-time} variable is a \lstinline!Real! variable without any prefix that is not assigned in a \lstinline!when!-clause.

>   A continuous-time variable \lstinline!vn! may have a non-vanishing time derivative (\lstinline!der(vn)<>0! possible) and may also change its value discontinuously at any time during transient analysis (see \cref{events-and-synchronization}).

Reviewer: This comment belongs to line 356: Introduce _continuous-time variable_ with `\firstuse` and `\index`.

Author: will add index, but unfortunately first-use is not versatile enough for this.

Reviewer: What's the problem with `\firstuse{continuous-time variable}`?

Author: The example below uses `\index{continuous time!variable}`; so if we later want `\firstuse` to map to an index-entry we would need to refine it, I thought it was easier to just use `\emph` for the time being.
There's also, separately, a need for cleaning up the index.

Reviewer: Looks good now.  Resolving.

\begin{itemize}
\item
  A variable \lstinline!vc! declared with \lstinline!constant!\indexinline{constant} prefix remains constant during transient analysis, with a value that is unaffected by the initialization problem.

>   This is called a constant, or constant variable.

Reviewer: I suppose this should be treated similar to introduced terminology for expression variability:
```suggestion
  This is called a \firstuse{constant}, or \firstuse{constant variable}\index{constant variable}\index{component variability!constant}.
```

\item
  A variable \lstinline!vc! declared with the \lstinline!parameter!\indexinline{parameter} prefix remains constant during transient analysis, with a value determined by the initialization problem.

>   This is called a parameter, or parameter variable.

Reviewer: I suppose this should be treated similar to introduced terminology for expression variability:
```suggestion
  This is called a \firstuse{parameter}, or \firstuse{parameter variable}\index{parameter variable}\index{component variability!parameter}.
```

> \item

Reviewer: Of course, not just parameter and constant component variabilities should be introduced with `\firstuse` and `\index`.  This comment is a reminder about the other ones.

Reviewer: This reminder is no longer needed.  Resolving.

\item
  A variable \lstinline!vc! declared with the \lstinline!parameter!\indexinline{parameter} prefix remains constant during transient analysis, with a value determined by the initialization problem.
\item

>   A \emph{discrete-time} variable \lstinline!vd! is a variable declared with the \lstinline!discrete!\indexinline{discrete} prefix, or a variable without any prefix that is not a continuous-time variable.

Author: That would not be correct, as Integer variables are not continuous-time variables even if not assigned in a when-clause.

>   A \emph{discrete-time} variable \lstinline!vd! is a variable declared with the \lstinline!discrete!\indexinline{discrete} prefix, or a variable without any prefix that is not a continuous-time variable.

Reviewer: True.  So how about avoiding the forward reference to continuous-time variable like this?
```suggestion
  A \emph{discrete-time} variable \lstinline!vd! is a variable that is discrete-valued (that is, not of \lstinline!Real! type) or assigned in a \lstinline!when!-clause.
  The \lstinline!discrete!\indexinline{discrete} prefix may be used to clarify that a continuous-valued (that is, of \lstinline!Real! type) variable is discrete-time.
```

I suppose we're assuming that a _variable_ can't be an external object?

>   A \emph{discrete-time} variable \lstinline!vd! is a variable declared with the \lstinline!discrete!\indexinline{discrete} prefix, or a variable without any prefix that is not a continuous-time variable.

Reviewer: I still think this could be a nice self-contained way of avoiding the forward reference.

>   A \emph{discrete-time} variable \lstinline!vd! is a variable declared with the \lstinline!discrete!\indexinline{discrete} prefix, or a variable without any prefix that is not a continuous-time variable.

Author: Will add a variant of this.

>   A \emph{discrete-time} variable \lstinline!vd! is a variable declared with the \lstinline!discrete!\indexinline{discrete} prefix, or a variable without any prefix that is not a continuous-time variable.

Reviewer: Looks good now.  Resolving.

  It has a vanishing time derivative between events.

>   Note that this is not the same as saying that \lstinline!der(vd)=0! almost everywhere, as the derivative is not even defined at the events, and it is not legal to apply \lstinline!der! to discrete-time variables as they are not continuous.

Reviewer: I find the motivation for not allowing `der` strange as it would apply just as fine to our continuous-time variables (which in general may even have discontinuities between events).  I doubt that there actually exist good reasons to not allow `der` for the discrete-time variables, and whatever we say should be harmonized with the definition of `der(expr)`:
> The time derivative of `expr`. If the expression `expr` is a scalar it needs to be a subtype of `Real`. The expression and all its time-varying subexpressions must be continuous and semi-differentiable.

Our discrete-time variables are clearly semi-differentiable, so the least we should do here is to not try to give any reason at all:
```suggestion
  Note that this is not the same as saying that \lstinline!der(vd)=0! almost everywhere, as the derivative is not even defined at the events.
  It is not allowed to apply \lstinline!der! to discrete-time variables.
```

However, this rule is so strange that we might as well consider dropping it right now.  From a technical point of view, applying `der` to a discrete-time variable is a breeze compared to applying it to some semi-differentiable `noEvent` (or `max`, for that matter) mess.  In particular, dealing with `der` applied to a discrete-time variable during index reduction would be a simple matter of removing the variable from the analysis, similar to a constant or parameter.

Reviewer: Sure, `max` with just two arguments isn't that bad.

My main point is still that as long as we allow `der` to be applied to continuous-time semi-differentiable expressions, it's just an arbitrary restriction to not allow it for discrete-time expressions (and variables).

  During transient analysis the variable can only change its value at event instants (see \cref{events-and-synchronization}).
\item

>   A \emph{continuous-time} variable is a \lstinline!Real! variable without any prefix that is not assigned in a \lstinline!when!-clause.

Reviewer: Alternatively, making the mutual exclusion of variable variabilities even more clear:
```suggestion
  A \emph{continuous-time} variable is a \lstinline!Real! variable without any prefix that is not a \emph{discrete-time} variable.
```

Author: I would prefer the current formulation, as it is more direct and the mutual exclusion seems to run the risk of circular dependency depending on how we define discrete-time variable.

Reviewer: Sure, we must ensure this is aligned with the definition of _discrete-time variable_.

Reviewer: Looks good now.  Resolving.

  A continuous-time variable \lstinline!vn! may have a non-vanishing time derivative (\lstinline!der(vn)<>0! possible) and may also change its value discontinuously at any time during transient analysis (see \cref{events-and-synchronization}).

>   If there are any discontinuities the variable is not differentiable.

Reviewer: What exactly is this statement about? Mathematical differentiability?  Applicability of `der`?

It would seem more relevant to state something directly related to applicability of `der`, if anything.

(Discussing this as part of this PR only makes sense in view of also trying to get the `der` part right regarding the discrete-time variables.)

Author: > What exactly is this statement about? Mathematical differentiability? Applicability of `der`?

Mathematical differentiability. The problem with applicability of `der` is that it is difficult/impossible to **statically** determine if there are any discontinuities or not.

Reviewer: OK, but then I think one would need to clarify why we are making statements about mathematical differentiability here, if this doesn't connect with any of the other concepts in the specification, such as `der`.

Reviewer: This still remains unresolved.

Author: This is now clarified; the point was to state that a continuous-time variable is general without any restrictions on time-evolution, so its trajectory may have parts that continuously change and also parts with discontinuities.

Reviewer: The clarification looks good, so then I'm just missing this part:
```suggestion
  Note that \lstinline!der(vn)! is not defined If there are any discontinuities, see \cref{modelica:der}.
```

Reviewer: Looks good now.  Resolving.

definition equations is given in \cref{variability-of-expressions}.

\begin{nonnormative}

> Note that for expressions discrete-time expressions include parameter expressions, whereas discrete-time variables do not include parameter variables.

Reviewer: Readability would be improved by removing the number of _expressions_:
```suggestion
Note that discrete-time \emph{expressions} include parameter expressions, whereas discrete-time \emph{variables} do not include parameter variables.
```
